### ISSUE TITLE
Provide guidelines about panic vs returning `Err`

### ISSUE THREAD

**Issue by athei** [paritytech/ink#1335]
We are inconsistent with panic vs. `Result` in our examples (they mostly panic). Biggest reason is that they were written before we had proper error handling (we only had panic to revert transactions back then).  I am not sure what our stance is on panic vs `Result` but we need to figure out what the best practices are.

We cannot just apply the guidelines from Rust where you use panics for bugs and `Result` for everything else because a contract differs from a normal program: In a normal Rust program having a potential panic is free but when it actually happens it is expensive because of unwinding. A `Result` is cheap in both cases but never free. So you want panics only for things that never happen (except in presence of a bug) and `Result` for everything else (runtime errors including invalid inputs).

In case of contracts a panic is always the cheapest option because we don't unwind. A panic is a single instruction (in addition to the branch) that immediately ends the execution without further costs. `Results` however are costly because data needs to be passed between contracts which is more overhead than passing between functions of a program. Additionally, we can argue that invalid input errors should never happen as there is another application (the Dapp) that can do the validation for free and should never let invalid input through. That puts them in the same category as bugs (just in another application of the stack) and panicing the way to go. Any other interaction that does not go through a dedicated Dapp (for example using `cargo-contract` or `Contracts UI`) is considered development and would involve the debug version of a contract where the panic message is visible and returning an error won't be necessary. An exception might be someone who interacts with a production contract through developers tools (and not a dedicated Dapp) out of curiosity or for learning purposes. But I would say this is a very niche use case we shouldn't consider that.

Considering all of that panicing seems to be the right way for input validation in general. This is consistent with the panic that can happen anyways when ink! SCALE decodes inputs which is implicit input validation from the point of view of an ink! user. 

This poses the question: When do we want to use an `Result` as an return type for a message given that it incurs extra costs even in the happy case: While wrapping a value you want to return into an `Option` is not that bad from an overhead cost but we also need to consider that having extensive `Error` enum could incur code size costs because all of the traits that are implemented on it.

I think there is only one very obvious case where we want to return a `Result`: When we expect that calling contract wants to process the result and behave differently based on it. But do we want to use it for other use cases, too?

The bottom line is that returning `Result` is the most flexible approach because it allows a calling contract do decide what to do with the error. The question we need to ask ourselves is: How low does the overhead need to be so that we would consider making `Result` the goto error handling strategy. If it is zero we need to say: panic whenever possible.
